File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Python/javonet/core/handler/GetInstanceMethodAsDelegateHandler.py
from javonet.core.handler.AbstractCommandHandler import *


class GetInstanceMethodAsDelegateHandler(AbstractCommandHandler):

    def __init__(self):
        self._required_parameters_count = 2

    def process(self, command):
        try:
            if len(command.payload) < self._required_parameters_count:
                raise Exception("InvokeInstanceMethod Parameters mismatch!")

            class_instance = command.payload[0]
            try:
                method = getattr(class_instance, command.payload[1])
            except AttributeError:
                methods = [method for method in dir(class_instance) if callable(getattr(class_instance, method))]
                message = "Method {} not found in class {}. Available methods:\n".format(command.payload[1],
                                                                                         class_instance.__class__.__name__)
                for method in methods:
                    message += "{}\n".format(method)
                raise AttributeError(message)

            return method
        except Exception as e:
            exc_type, exc_value = type(e), e
            new_exc = exc_type(exc_value).with_traceback(e.__traceback__)
            raise new_exc from None

File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Nodejs/lib/core/handler/GetInstanceMethodAsDelegateHandler.js
const AbstractHandler = require('./AbstractHandler')

/**
 * Handles retrieving an instance method as a delegate.
 */
class GetInstanceMethodAsDelegateHandler extends AbstractHandler {
    constructor() {
        super()
        /** @type {Function|null} */
        this.method = null
        /** @type {number} */
        this.requiredParametersCount = 2
        /** @type {Array<*>} */
        this.args = []
        /** @type {Object|null} */
        this.instance = null
    }

    /**
     * Processes a command to retrieve an instance method as a delegate.
     * @param {Object} command - The command containing payload data.
     * @param {Array<*>} command.payload - The payload containing instance, method name, and arguments.
     * @returns {Function} The delegate for the instance method.
     * @throws {Error} If the parameters mismatch or the method cannot be found.
     */
    process(command) {
        if (command.payload.length < this.requiredParametersCount) {
            throw new Error(`${this.constructor.name} parameters mismatch`)
        }

        this.instance = command.payload[0]
        const methodName = command.payload[1]
        this.args = command.payload.length > 2 ? command.payload.slice(2) : []

        // Find the method on the instance
        this.method = this.getMethod(this.instance, methodName)
        if (!this.method) {
            throw this.createMethodNotFoundError(this.instance, methodName)
        }

        // Create a delegate from the method
        const methodDelegate = this.method.bind(this.instance)
        return methodDelegate
    }

    /**
     * Retrieves the method from the type.
     * @param {Object} type - The class or constructor to search for the method.
     * @param {string} methodName - The name of the method.
     * @returns {Function|null} The found method or null if not found.
     */
    getMethod(type, methodName) {
        const method = type[methodName]
        return typeof method === 'function' ? method : null
    }

    /**
     * Creates an error message when the method is not found.
     * @param {Object} type - The class or constructor.
     * @param {string} methodName - The method name.
     * @returns {Error} The error with detailed message.
     */
    createMethodNotFoundError(type, methodName) {
        const methods = Object.getOwnPropertyNames(type.prototype).filter(
            (key) => typeof type.prototype[key] === 'function'
        )
        const availableMethods = methods.map((name) => `${name}()`)
        const message = `Method ${methodName} not found in class ${type.name}. Available instance public methods:\n${availableMethods.join('\n')}`
        return new Error(message)
    }
}

module.exports = new GetInstanceMethodAsDelegateHandler()
